### ISSUE TITLE
Implicit Promotion - Parallel only for statically-proven 'pure' functions

### ISSUE THREAD

**Issue by LouisJenkinsCS** [chapel-lang/chapel#12803]
I like the ability to implicitly vectorize functions on vector-like arguments (such as arrays, ranges, domains, etc.), but it would be nice if the compiler would perform this serially rather than in parallel when the function is a non-pure one. Even if the function has synchronization, it likely will not be any faster than serial execution, and worst-case you get undefined behavior.

```chpl
// Pure; should parallelize
proc f(x : int) {
   return x + 1;
}
// Impure; should not parallelize
var state : int;
proc g(x : int) {
   state += x;
}
// Impure but access is through atomics; should parallelize
var state : atomic int;
proc h(x : int) {
   state.add(x);
}
```

**Comment by LouisJenkinsCS** [comment]
What if you can force serial promotion to parallel by using square-brackets? `g([A])` can force it to be parallel, which requires the user to _know_ that implicit promotion is occurring. 

Also side note: from some feedback from students using the language, the implicit promotion is very confusing to beginners and I got a lot of "How would I even know the compiler was going to do that" kind of disbelief when I explained the issue.

**Comment by bradcray** [comment]
I understand the motivation for this request, but feel nervous about making Chapel semantics be governed by whether or not a compiler can prove something since one compiler's ability to do so could differ from another's, so am thinking about other ways of minimizing these issues.  

I think that the impure case in your example is a variation on a broader concern that could occur in any parallel loop that refers to a module-scope variable:

```chapel
var state : int;

proc f() {
  forall i in 1..10000 do
    state += i;
}

f();
```

However, the code above generates an error while the promoted case does not, which seems like an inconsistency:

```chapel
proc g(x: int) {
  state += x;
}

g(1..10000);
```

@vasslitvinov, do you have any insight into why these cases would differ and/or what it would take to make them more consistent?  (I think @mppf may have done some work in a related area as well recently regarding promotions and safety, so I'm tagging him as well)

My other thought here (which is something I've wanted for awhile myself) is to add some sort of `--report-implicit-parallelism` flag so that once you're attuned to such things, you'd have an easy way to determine where they're occurring (e.g., it's easy to grep a program for instances of begin, cobein, coforall, and forall, but would be nice to have a flag to help find implicit parallelism as well).  And at times like this, I've also imagined a `--newbie` flag that would be recommended for users who are new to the language and help point out things that they might not pick up on (e.g., "are you aware that you're promoting a function here?").

I'd also be curious how your students ran into this problem.  E.g., Did they think they were calling a routine that took an array when it actually only took a scalar?  Or something else?

**Comment by vasslitvinov** [comment]
> why these cases would differ

If we make the promotion in g(1..10000) explicit, the two cases are:

```chpl
var state = 0;
proc g(x:int) { state += x; }

forall i in 1..10000 do  state += i;     // case 1: forall loop

forall i in 1..10000 do  g(i);           // case 2: promotion
```

The difference is whether the outer variable `state` is accessed within the lexical scope of the forall loop or from somewhere within the closure of the call graph.

When we introduced the "in intent" feature, we explicitly decided to handle the former and not the latter. From the commit message of [r21940](https://github.com/chapel-lang/chapel/commit/0a0757ce5a) :

> Do not capture the "closure" variables, i.e., the variables in the
> dynamic scope of a task construct. This is to keep capture manageable,
> so that all variables subject to capture can be obtained by examining
> just the lexical scope. The set of closure variables can be hard to
> determine, unweildy to implement and reason about, it is unclear
> what to do with extern functions, etc.

If anybody would like to revisit this decision, please open a new issue.

**Comment by vasslitvinov** [comment]
> the implicit promotion is very confusing to beginners 

@LouisJenkinsCS - in your experience, when a beginner writes code that can be reduced to something like:

```chpl
proc foo(x:int) { <whatever> }
foo(1..100);
```

what do they intuitively expect to happen?

One thought is to create a flag that tells the compiler "I am a beginner, go easy on me." With the flag on, the compiler would issue a "this feature is not for beginners" error for the less-intuitive features. Although it is surely non-trivial to implement this flag well.

**Comment by LouisJenkinsCS** [comment]
It was a logical error on their part; they "accidentally" passed an array to a scalar argument; it was more that they weren't aware that they made the error. Essentially they were looking elsewhere for a bug in their code, and it was difficult to find the issue given the implicit type inference. I forget the specifics though.

**Comment by LouisJenkinsCS** [comment]
To recreate it somewhat (I genuinely do not remember what the original context was), see [this example](https://tio.run/##dY8xT8MwEIVn/Cve1kQVKWEk6hAxMcNWdQjOObGU2tZxbRMQvz3YVZUiIcY73fved7pvAg3zvNng1R8I5ui0WO9wttKjcaAxDFZbgUyBIH0jeEFLxjpqi6JQgb2GyRyeYJ3k@FJ3THJkB4c1ykp9K6Ui/S3FrTPETC3eJ2h/CHYgTpB/2peb@yV4s5Cepj8iXTb@VjBxvBicGkYdFXdlUZQP@6SKLS5Dldqfr0U3wysh9SV6AsRVDHVZvSv38bc6rxCjxnPno46P2ZZGrOoVemJanjqRFs/2M0J9IG7Sgx@JmXjrLR4rdWYrNLgsbvJqnn8A)

```chpl
// Some function with an explicit type that I defined...
proc f(n : int) {
	return n + 1;
}


// Type inferred by compiler...
// Some function with a compiler-inferred type that they defined...
proc g(x) {
	return f(x);
}

var A : [1..10] int = 1..10;
// Compiler inferred return type...
var ret = g(A[1] + A); // forgot to index 'A' here
// Some vectorized operations...
ret += 2;
writeln(ret);
```

In essence: Imagine that somewhere down the pipeline, you accidentally passed a vector as a scalar and then continued using the return type as a scalar via Chapel's Array vector operations. If you eventually use it in an illegal manner that results in compiler error, it results in confusion: "Why do I have an `_ir_chpl_*` type in the error message?" and if it does work, you get unexpected behavior (although it is admittedly the newcomer's fault).

I'm not saying that we do-away with implicit promotion either, just that in the worst-case it shouldn't result in undefined behavior because they pass in an impure function (I.E by making it sequential when the compiler cannot prove that parallelism is safe).

**Comment by LouisJenkinsCS** [comment]
Either way, the side-note wasn't meant to diverge this issue from discussing sequential implicit promotion for impure functions.
